File: projects/09/clojure/src/Temp.clj
;; Temp.clj is a module for Temp
;; name: Daniela Contreras
;; Worked with Germaine Hounakey
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Create a record/struct type named Temp
(defrecord Temp [myDegrees myScale])

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; make-Temp constructs a Temp object from a double and char.
;; Receive: degrees, a double; scale, a char
;; Return: the Temp (degrees scale).
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn make-Temp [degrees scale]
  (->Temp degrees scale)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; isValid checks temperature's validity
;; Receive: degree, a double; scale, a char
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn isValid [degree scale]
  (if (or (and (= scale (or "K" "k")) (>= degree 0.0)    )
          (and (= scale (or "C" "c")) (>= degree -273.15))
          (and (= scale (or "F" "f")) (>= degree -459.67)))
    true
    false
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; getDegrees extracts the degrees of a Temp object.
;; Receive: aTemp, a Temp.
;; Return: myDegrees double in aTemp
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn getDegrees [aTemp]
  (:myDegrees aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; getScale extracts the scale of a Temp object.
;; Receive: aTemp, a Temp.
;; Return: myScale char in aTemp
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn getScale [aTemp]
  (:myScale aTemp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; writeTemp prints temperature
;; Receive: aTemp, a Temp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn writeTemp [aTemp]
  (print (:myDegrees aTemp) (:myScale aTemp))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; readTemp prints temperature
;; Receive: aTemp, a Temp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn readTemp []
  (print "Enter Degrees and Scale: ") (flush)
  (def newDegree (double (read)))
  (def newScale (str (.charAt (str (read)) 0)))
  (if (isValid newDegree newScale)
    (make-Temp newDegree newScale)
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toF converts Temp to Fahrenheit
;; Receive: aTemp, a Temp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toF [aTemp]
  (def newDegree (:myDegrees aTemp))
  (if (= (:myScale aTemp) (or "C" "c"))
    (def newDegree (+ (* (/ 9 5) (:myDegrees aTemp)) 32)))
  (if (= (:myScale aTemp) (or "K" "k"))
      (def newDegree (+ (* (/ 9 5) (- (:myDegrees aTemp) 273.15)) 32)))
  (make-Temp newDegree "F")
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toC converts Temp to Celsius
;; Receive: aTemp, a Temp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toC [aTemp]
  (def newDegree (:myDegrees aTemp))
  (if (= (:myScale aTemp) (or "F" "f"))
    (def newDegree (- (:myDegrees aTemp) 459.66999999999996)))
  (if (= (:myScale aTemp) (or "K" "k"))
    (def newDegree (- (:myDegrees aTemp) 273.15)))
  (make-Temp newDegree "C")
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toK converts Temp to Kelvin
;; Receive: aTemp, a Temp.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toK [aTemp]
  (def newDegree (:myDegrees aTemp))
  (if (= (:myScale aTemp) (or "F" "f"))
    (def newDegree (+ (* (/ 5 9) (- (:myDegrees aTemp) 32)) 273.15)))
  (if (= (:myScale aTemp) (or "C" "c"))
    (def newDegree (+ (:myDegrees aTemp) 273.15)))
  (make-Temp newDegree "K")
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; isEqual checks temperature's equality scales
;; Receive: Left, a Temp; Right, a Temp
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn isEqual [Left Right]
  (def LeftCopy (toK Left))
  (def RightCopy (toK Right))
  (= (:myDegrees LeftCopy) (:myDegrees (toK RightCopy)))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; isLessThan checks temperature's lessThan across scales
;; Receive: Left, a Temp; Right, a Temp
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn isLessThan [Left Right]
  (< (:myDegrees (toF Left)) (:myDegrees (toF Right)))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; raiseTemp increases temperature
;; Receive: aTemp, a Temp; degree, a double
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn raiseTemp [aTemp degree]
  (if (isValid (+ (:myDegrees aTemp) degree) (:myScale aTemp))
    (make-Temp (+ (:myDegrees aTemp) degree) (:myScale aTemp))
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; lowerTemp decreases temperature
;; Receive: aTemp, a Temp; degree, a double
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn lowerTemp [aTemp degree]
  (if (isValid (- (:myDegrees aTemp) degree) (:myScale aTemp))
    (make-Temp (- (:myDegrees aTemp) degree) (:myScale aTemp))
  )
)

File: projects/09/ruby/Temp.rb
#######################################################################################################################
#### Name: Daniela Contreras
#### Date: 03/19/2022
### Temp.rb is a module containing the class Temperature
###
module Temperatures
    class Temp
      # Initializes Temp
      def initialize(degrees, scale)
        if isValid(degrees, scale) then
        @degree = degrees
        @scale = scale
        end
      end
  
      # Accessors for Temp
      attr_reader :degree, :scale
  
      #######################################################################
      # readTemperature: reads in degrees and its scale
      # Output: user sets new value
      #######################################################################
      def readTemperature
        print "Enter Degrees and Scale: "
        temperature = gets
        anArray = temperature.split(" ")
        initialize( anArray[0].to_f, anArray[1][0,1] )
      end
  
      #######################################################################
      # writeTemperature: prints the temperature to the user/screen
      #######################################################################
      def writeTemperature
         print @degree.round(2).to_s + " " + @scale.to_s
      end
  
      #######################################################################
      # toF: converts Temperature to Fahrenheit
      #######################################################################
      def toF
        if @scale.upcase == 'K' then
            @degree = (@degree-273.15)*9/5+32;
        elsif @scale.upcase == 'C' then
            @degree = @degree*9/5+32;
        end
        @scale = 'F';
      end
  
      #######################################################################
      # toC: converts Temp to Celsius
      #######################################################################
      def toC
        if @scale.upcase == 'K' then
            @degree = @degree-273.15;
        elsif @scale.upcase == 'F' then
            @degree = (@degree-32)*5/9;
        end
        @scale = 'C';
      end
      #######################################################################
      # toK: converts Temp to Kelvin
      #######################################################################
      def toK
        if @scale.upcase == 'C' then
          @degree = @degree + 273.15;
        elsif @scale.upcase == 'F' then
            @degree = (@degree-32)*5/9+273.15;
        end
        @scale = 'K';
      end
  
      #######################################################################
      # toScale: converts Temp to scale of choice
      #######################################################################
      def toScale(scale)
        if    scale.upcase == "F" then
          toF
        elsif scale.upcase == "C" then
          toC
        elsif scale.upcase == "K" then
          toK
        end
      end
  
      #######################################################################
      # raiseTemp: increases Temp by degree
      # Receive: aTemp, a Temp
      #######################################################################
      def raiseTemp(degree)
        if isValid(@degree+degree, @scale) then
        @degree = @degree + degree
        end
      end
  
      #######################################################################
      # lowerTemp: decreases Temp by degree
      # Receive: aTemp, a Temp
      #######################################################################
      def lowerTemp(degrees)
        if isValid(@degree-degrees, @scale) then
              @degree = @degree - degrees
        end
      end
  
      #######################################################################
      # isEqual: checks the equality of Temperatures across scales
      #######################################################################
      def isEqual(temp)
         origScale1 = @scale
         origScale2 = temp.scale
         temp.toK
         toK
         boolTest = (@degree == temp.degree)
         toScale(origScale1)
         temp.toScale(origScale2)
         return boolTest
      end
  
      #######################################################################
      # isLessThan: checks lessThan of Temperatures 
      #######################################################################
      def isLessThan(temp)
         origScale2 = temp.scale
         origScale1 = @scale
         temp.toK
         toK
         boolTest = (@degree < temp.degree)
         toScale(origScale1)
         temp.toScale(origScale2)
         return boolTest
      end
      #######################################################################
      # isValid: checks Temperatures
      #######################################################################
      def isValid(degree, scale)
        if    scale.upcase == 'K' and degree >= 0.00    then
          return true 
        elsif scale.upcase == 'C' and degree >= -273.15 then
    return true
        elsif scale.upcase == 'F' and degree >= -459.67 then
    return true
        else
    puts "Error: isValid(), Invalid Temperature."
    return false
        end
      end
    end
  end